### PR TITLE
Make standalone-bundle OSGi compatible

### LINKED ISSUE #269: Make HiveMQ Client OSGi compliant
Hello,

could you please make the client library OSGi compliant be generating a full MANIFEST.MF file with the correct IMPORT and EXPORT packages statements.

This would allow users to deploy the client into an OSGi container (Felix, Equinox, Karaf ...) without any further actions required.

maven has the required plugins to generate valid OSGi bundles. The best and most reliable solution would probably be to use the bnd(tools) plugin for maven.

https://enroute.osgi.org/about/115-bnd-plugins.html

Many thanks in advance

### PR DESCRIPTION
**Pull request by J-N-K**
Resolves #269 

The shaded bundle does not benefit from OSGi's modular concept, therefore I did not touch that.

Signed-off-by: Jan N. Klug <jan.n.klug@rub.de>

### CODE REVIEW

**Comment by J-N-K** [comment]
@SgtSilvio, can you point me to the CLA?

**Comment by SgtSilvio** [comment]
@J-N-K about the CLA: https://github.com/hivemq/hivemq-community/blob/master/CONTRIBUTING.adoc

File: build.gradle [hunk 1]
```diff
@@ -1,3 +1,12 @@
+buildscript {
+    repositories {
+        mavenCentral()
+    }
+    dependencies {
+        classpath 'biz.aQute.bnd:biz.aQute.bnd.gradle:4.2.0'
```
Comment (SgtSilvio) [inline build.gradle]: I think the plugin can also be applied in the `plugins` block
`id 'biz.aQute.bnd.builder' version '4.2.0'`

File: bnd.bnd [hunk 2]
```diff
@@ -0,0 +1,10 @@
+Bundle-Description: HiveMQ Java MQTT Client 
+Bundle-License: Apache 2.0 License
+Export-Package: \
+  com.hivemq.client.mqtt.*, \
+  com.hivemq.client.mqtt.rx.*, \
+  com.hivemq.client.mqtt.rx.reactivestreams.*, \
```
Comment (SgtSilvio) [inline bnd.bnd]: I am not a OSGI expert, but isn't `com.hivemq.client.rx.reactivestreams.*` covered by `com.hivemq.client.rx.*`?

File: bnd.bnd [hunk 3]
```diff
@@ -0,0 +1,10 @@
+Bundle-Description: HiveMQ Java MQTT Client 
+Bundle-License: Apache 2.0 License
+Export-Package: \
+  com.hivemq.client.mqtt.*, \
+  com.hivemq.client.mqtt.rx.*, \
```
Comment (SgtSilvio) [inline bnd.bnd]:
```suggestion
  com.hivemq.client.rx.*, \
```

File: bnd.bnd [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  com.hivemq.client.mqtt.rx.reactivestreams.*, \
+  com.hivemq.client.mqtt.util.*
```
Comment (SgtSilvio) [inline bnd.bnd]:
```suggestion
  com.hivemq.client.util.*
```

File: bnd.bnd [hunk 5]
```diff
@@ -0,0 +1,10 @@
+Bundle-Description: HiveMQ Java MQTT Client 
+Bundle-License: Apache 2.0 License
+Export-Package: \
```
Comment (SgtSilvio) [inline bnd.bnd]: `com.hivemq.client.annotations.*` is missing

File: bnd.bnd [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+  com.hivemq.client.mqtt.*, \
+  com.hivemq.client.mqtt.rx.*, \
+  com.hivemq.client.mqtt.rx.reactivestreams.*, \
+  com.hivemq.client.mqtt.util.*
+Import-Package: \
+  *
```
Comment (SgtSilvio) [inline bnd.bnd]: We should import the dependencies specifically.
Can the gradle plugin generate this automatically from the dependencies?

File: bnd.bnd [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+-consumer-policy: ${range;[==,=+)}
```
Comment (SgtSilvio) [inline bnd.bnd]: Why is this necessary?

Comment (J-N-K) [reply bnd.bnd]: Standard policy is `[==,+)` which would be (for netty) `[4.1,5)`, i.e. 4.1 and above, below 5. Are we sure that 4.2 (or 4.9) is fully bckward compatible? If so, we can remove that.

File: bnd.bnd (hunk 6 above)
Comment (J-N-K) [reply bnd.bnd]: This can be removed at all, it's a left-over and bnd's standard. bnd only imports needed packages anyway. I tried to exclude some imports (make them optional) before but that did not work (I thought I could make rxjava optional, so you don't have to provide that if you don't use the trx-API, but it's still needed internally.

**Comment by SgtSilvio** [comment]
@cla-bot check

File: build.gradle (hunk 1 above)
Comment (J-N-K) [reply build.gradle]: I'm not very familiar with gradle (to be honest, I've never worked with it), but it won't work for my local build if I demove that section. I applied it in the plugins-block.

**Comment by J-N-K** [comment]
One question (this is not mandatory for an OSGi-bundle but IMO useful):

In openHAB we add the followong header. Shall I add something like that? And what should be the content?

```
Bundle-Developers: openhab;email="info@openhab.org";name=openHAB;organ
 ization=openHAB;organizationUrl="https://www.openHAB.org"
```

File: bnd.bnd [hunk 8]
```diff
@@ -0,0 +1,16 @@
+Automatic-Module-Name: ${def;bsn}
+Bundle-Description: HiveMQ Java MQTT Client
```
Comment (SgtSilvio) [inline bnd.bnd]: We have all these values (name, license, scm) as variables in the gradle build script.
We could use them if we declare the bnd rules in the gradle build file: https://github.com/bndtools/bnd/blob/master/biz.aQute.bnd.gradle/README.md#instructing-bnd-on-how-to-build-your-bundle

File: build.gradle (hunk 1 above)
Comment (J-N-K) [reply build.gradle]: This is due to https://github.com/bndtools/bnd/issues/3174. After 4.3.0 is released the block can be removed again.

File: build.gradle [hunk 9]
```diff
@@ -90,6 +100,18 @@ apply from: 'japicc.gradle'
 
 import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar
 
+jar {
+    bnd('Automatic-Module-Name' : jar.baseName,
+        'Bundle-Name': readableName,
+        'Bundle-Description': rootProject.description,
+        'Bundle-Vendor': 'dc-square and the HiveMQ MQTT Client Project',
+        'Bundle-License': licenseShortName + ';description="' + licenseReadableName + '"link="' + licenseUrl + '"',
```
Comment (SgtSilvio) [inline build.gradle]:
```suggestion
        'Bundle-License': licenseShortName + ';description="' + licenseReadableName + '";link="' + licenseUrl + '"',
```

**Comment by SgtSilvio** [comment]
I changed the base branch as I want to merge your PR but we still have to decide on the automatic-module-name and the versioning policy before merging it into develop.

**Comment by SgtSilvio** [comment]
@J-N-K did you try if the OSGi bundle works? I am wondering because the following two packages are listed as imports: `dagger` and `dagger.internal`, but without a version and dagger actually does not have an OSGi bundle.

**Comment by J-N-K** [comment]
I‘ll check the latest version this afternoon. Non-OSGi-bundle imports are not a problem. The consumer can always wrap the bundle himself (there is an explanation in the bnd-documentation). Some OSGi containers like Karaf can do that on-the-fly (if configured to do so). The difference to a real OSGi-bundle is the missing metadata and (more important) all dependencies are marked „optional“, which is most probably wrong.

**Comment by J-N-K** [comment]
The two most-common Automatic-Module-Name-Headers would be the bundle symbolic name (what we have now) or groupID.bundleName, up to you.

I can also confirm that ithe bundle works in my setup.

File: bnd.bnd (hunk 7 above)
Comment (J-N-K) [reply bnd.bnd]: I checked https://netty.io/news/2013/08/27/semantic-versioning.html and I think that the more restrictive policy I have chosen is correct (at least for netty) because different minor versions are not guaranteed to be binary compatible (which is what we need here).

I can add a version for the dagger import if you want.

**Comment by SgtSilvio** [comment]
The Automatic-Module-Name is used for the Java Module System and should be the reverse DNS and reflect the package names (https://blog.joda.org/2017/04/java-se-9-jpms-module-naming.html)
So it should be `com.hivemq.client` or `com.hivemq.client.mqtt`.

File: bnd.bnd (hunk 7 above)
Comment (SgtSilvio) [reply bnd.bnd]: Is there any advantage of adding a version for dagger?

**Comment by SgtSilvio** [comment]
Decision:
Automatic-Module-Name and Bundle-SymbolicName: com.hivemq.client.mqtt
Bundle-Name: hivemq-mqtt-client (project.name)
Bundle-Vendor: HiveMQ and the HiveMQ Community

**Comment by SgtSilvio** [comment]
I will merge your PR and apply the decisions.

**Comment by SgtSilvio** [comment]
@J-N-K Thanks again for your contribution. PR #319 will merge your changes to the develop branch.

**Comment by SgtSilvio** [comment]
@J-N-K your contribution was released as part of version 1.1.2: https://github.com/hivemq/hivemq-mqtt-client/releases/tag/v1.1.2
